### PR TITLE
Added trait to support Codeception tests

### PR DESCRIPTION
**Pull request by pbatroff**
Overview
----------------------------------------
[Codeception](http://codeception.com/) is a PHP-based testing framework. The upshot of this particular tool is its breadth -- the test-runner/code-generator/documentation present three levels of testing (e2e/acceptance, functional/integration, and unit).

However, it does not have any built-in support for interacting with a CiviCRM installation, and it would be redundant for many extension authors to write their own Codeception helper integration functions. This patch defines a utility class which can be incorporated into other downstream tests.

We (Philipp B, Michael D, and Tim O) initiated this project at the sprint in Devon with the aim of supporting end-to-end extension testing. 

Before
----------------------------------------
No Codeception traits available

After
----------------------------------------
Codeception tests for extensions can now interact with CiviCRM if they use this trait. 

Technical Details
----------------------------------------

* Codeception's acceptance tests often use the helper `$I->amOnPage(...)`. However, the URL differences of Drupal/Backdrop/Joomla/WordPress make this difficult to use. The trait adds a similar function `$I->amOnRoute(...)`.
* The trait provides a CMS-independent login interface: `$I->login($user,$pass)`, `$I->loginAsAdmin()`, `$I->loginAsDemo()`.
* The main `login($user,$pass)` function needs to be adapted for different CMS's. This is implemented for D7, and it has placeholders for other CMSs.
* There's a parallel [update for `civix`](https://github.com/totten/civix/pull/111) to initialize an extension's Codeception configuration. It supports booting Civi and loading the `CiviAcceptanceTesterTrait` 

Comments
----------------------------------------

### CODE REVIEW

**Comment by totten** [comment]
jenkins, add to whitelist

File: Civi/Test/CiviCodeceptionTrait.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  public function amOnRoute($page) {
+
+    $params = explode('?', $page);
+    if (empty($params[1])) {
+      $newPage = \CRM_Utils_System::url($page, NULL, FALSE, NULL, FALSE);
+    }
+    else {
+      $newPage = \CRM_Utils_System::url($params[0], $params[1], FALSE, NULL, FALSE);
+    }
+    error_log($newPage);
```
Comment (totten) [inline Civi/Test/CiviCodeceptionTrait.php]: Old debug code?

File: Civi/Test/CiviCodeceptionTrait.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Trait CiviCodeceptionTrait
+ * Trait for common functions used for Codeception End-to-End Testing framework
+ * @package Civi\Test
+ */
+trait CiviCodeceptionTrait {
+
+  /**
+   * Parse Parameters, and create generic Civi URl
+   * @param $page
```
Comment (totten) [inline Civi/Test/CiviCodeceptionTrait.php]: The `@param` docbocks should include a little more detail. Example:

```php
  * @ param string $page
  *    The path and parameters of a CiviCRM page. Ex: "civicrm/dashboard?reset=1".
```

The same would apply to other `@param`s.

**Comment by eileenmcnaughton** [comment]
@totten this looks low to no risk & you have reviewed & fixes have been made in response - good to merge?

**Comment by eileenmcnaughton** [comment]
@totten ^^

File: Civi/Test/CiviCodeceptionTrait.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  /**
+   * Dispatcher for lonin to supported plattforms
+   * @param $username
+   *   CiviCRM username for the login
+   * @param $password
+   *   CiviCRM password for the login
+   */
+  public function login($username, $password) {
+    $config = CRM_Core_Config::singleton();
+    switch ($config->userframework) {
```
Comment (totten) [inline Civi/Test/CiviCodeceptionTrait.php]: You can simplify this (removing the hard-coded names) with a snippet like:

```php
  $handler = array($this, 'loginTo' . $config->userFramework);
  if (is_callable($handler)) {
    call_user_func($handler, $username, $password);
  } else {
    throw new FooException(...);
}
```

File: Civi/Test/CiviCodeceptionTrait.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
+<?php
+namespace Civi\Test;
+
+/**
+ * Trait CiviCodeceptionTrait
+ * Trait for common functions used for Codeception End-to-End Testing framework
+ * @package Civi\Test
+ */
+trait CiviCodeceptionTrait {
```
Comment (totten) [inline Civi/Test/CiviCodeceptionTrait.php]: We discussed this a little face-to-face -- it'd be better to rename this `\Civi\Test\Codeception\CiviAcceptanceTesterTrait`. This naming convention makes it easier to differentiate the helper functions/implementations for acceptance/functional/unit tests.

**Comment by totten** [comment]
@pbatroff I rewrote the description.

**Comment by totten** [comment]
(*CiviCRM Review Template WORD-1.0*)



* ([`r-jira`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-jira)) __Waived__: Only affects developers. Expecting [r-explain](https://github.com/civicrm/civicrm-dev-docs/pull/474) to come in.
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __Pass__: There's one test failure, `CRM_Utils_GeocodeTest::testGeocodeMethodOff` ("CiviCRM_API3_Exception: Aborting batch geocoding. Hit the over query limit on geocoder."). This is a false-negative. 
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Pass__
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __WIP__: Needs dev docs. I'm hoping to do this during the remaining bit of the sprint and don't want to block merging.
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Pass(ish)__: Technically, one might want to write test-coverage for the test-utilities. But we're not actually using Codeception in `civicrm-core`, so... it's a little awkward to instrument the test.
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __Pass__: Generated configuration and test scripts with `civix` and `codecept` (respectively). The ran successfully. Also tried https://gist.github.com/totten/cbd9cdfba5f735d3c6c70ae22bdffada
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __Pass__: Doesn't affect user-experience.
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Pass__: Adds a new file which only loads on demand (as part of some test-suites) -- and doesn't affect existing integrations/extensions/suites.

**Pull request merged**